working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+`Ext1` will treat the underlying object according to the discipline
+enforced by `Ext2`, and then add its own layer of extra discipline.*
+
+When `E` is a non-protected extension type, a type test `o is E` and a type
+check `o as E` can be performed. Such checks performed on a local variable
+can promote the variable to the extension type using the normal rules for
+type promotion.
+
+*Protected extension types are introduced in a section below, and they have
+a more strict rule for type tests and type casts.*
+
+*Compared to the existing extension methods feature, there is no change to
+the type of `this` in the body of an extension type _E_: It is the on-type
+of _E_. Similarly, extension methods of _E_ invoked in the body of _E_ are
+subject to the same treatment as previously, which means that extension
+methods of the enclosing extension type can be invoked implicitly, and
+extension methods are given higher priority than instance methods on `this`
+when `this` is implicit.*
+
+
+### Prevent implicit invocations: Keyword 'type'
+
+This section specifies the effect of including the keyword `type` in the
+declaration of an extension type.
+
+*A reminder, to set the scene: An implicit extension member invocation is
+an invocation of the form `e.m()` (or similar for getter/setter/operator
+invocations) which invokes a member of an extension `E`, but where the
+static type of `e` is a type `T` that matches the on-type of `E`. This is
+the typical usage for the extension methods mechanism that Dart has had
+since version 2.6, and it is available with no changes for an extension
+declaration that does not include the keyword `type`. The other kinds of
+invocations of extension members are explicit extension member invocations
+(like `Ext(e).m()`) and typed extension member invocation (like `e.m()`
+where the static type of `e` is `Ext`).*
+
+Let `E` be an extension declaration such that the keyword `extension` is
+followed by `type`. In this case, `E` is not applicable for an implicit
+extension method invocation.
+
+*Such an extension type is not applicable to implicit invocations where a
+plain extension would be applicable. The extension otherwise works the
+same as an extension without the `type` modifier. For example:*
+
+```dart
+extension type Age on int {
+  Age get next => this + 1;
+}
+
+void main() {
+  int i = 42;
+  i.next; // Error, no such method.
+  Age age = 42;
+  age.next; // OK.
+}
+```
+
+When the keyword `extension` is followed by `type` in an extension
+declaration, we say that it is an _explicit_ extension type declaration,
+and that it introduces an _explicit_ extension type.
+
+*This terminology is motivated by two things: (1) The extension type "is
+not implicit", because it doesn't allow for implicit extension method
+invocations. (2) The extension type explicitly says `type`, and it must
+also be used explicitly as a type.*
+
+Let `E` be an explicit extension type declaration. It is not an error to
```

Comment:
Hmm, this probably needs a discussion issue.  I think this will be controversial.

Reply:
Certainly. Created https://github.com/dart-lang/language/issues/1462 to hold the discussion.